### DISCUSSION TITLE
Issues with MQTT not receiving messages after ~25 seconds

### DISCUSSION THREAD

**Discussion by hstriegel** [micropython/micropython#10795] [RP2040 / Pico] [answered]
As the title suggests, MQTT subscription works until about 25 seconds, where it then doesn't receive any requests. I used the hivemq websocket to try to get it to work, when I ran into this issue. Any help would be appreciated!

Here is the Code:

```py
import network
import time
from machine import Pin
from umqtt.simple import MQTTClient

wlan = network.WLAN(network.STA_IF)
wlan.active(True)
wlan.connect("*****",security=0)
time.sleep(5)
print(wlan.isconnected())

LED = Pin("LED", Pin.OUT)

mqtt_server = 'broker.hivemq.com'
client_id = 'client'
topic_sub = b'makesurethisworks'


def sub_cb(topic, msg):
    print("New message on topic {}".format(topic.decode('utf-8')))
    msg = msg.decode('utf-8')
    print(msg)
    if msg == "on":
        LED.on()
    elif msg == "off":
        LED.off()

def mqtt_connect():
    client = MQTTClient(client_id, mqtt_server, keepalive=60)
    client.set_callback(sub_cb)
    client.connect()
    print('Connected to %s MQTT Broker'%(mqtt_server))
    return client

def reconnect():
    print('Failed to connect to MQTT Broker. Reconnecting...')
    time.sleep(5)
    machine.reset()
    
try:
    client = mqtt_connect()
except OSError as e:
    reconnect()
client.subscribe(topic_sub, 0)
while True:
    client.wait_msg()
```

**Comment by davefes** [comment]
Are you using their free service?

**Reply by hstriegel** [reply]
Yes

**Reply by davefes** [reply]
I had a go at MQTT a few months ago and as I recall the free HiveMQ only saved your message for a short period.  I wanted a device to check-in for messages an hour to 24 hours later, but didn't find a free one that would do what I wanted.

You could test using the Mosquitto broker but they suggest that you don't use it for more than the odd test.  If you find something I'd be interested.  Have you looked at Telegram?

**Comment by peterhinch** [comment]
The `umqtt.simple` driver has a number of limitations, notably it doesn't keep the link open with MQTT ping requests. Users have had success with Hive MQ using [asynchronous MQTT](https://github.com/peterhinch/micropython-mqtt/blob/master/mqtt_as/README.md) and the doc has a section on use with that broker.

**Comment by TaylorJayM** [comment]
Under your while true loop change client.wait_msg() to:

```
try:
   client.check_msg()
except:
   try:
      client.connect()
      client.subscribe(topic_sub)
   except:
      print('MQTT Reconnect Failed')
```

**Comment by hstriegel** [comment, marked as answer]
I eventually figured it out that my college campuses wifi did not allow Raspberry Pis and so that was the reason that I got disconnected.
